src/System.Text.ValueBuilder/System/Text/ValueStringBuilder.cs

```diff
@@ -0,0 +1,277 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Text
+{
+    // *** This file is a simple copy from CoreFX/CoreCLR (made public)
+
+    public unsafe ref partial struct ValueStringBuilder
+    {
+        private char[] _arrayToReturnToPool;
+        private Span<char> _chars;
+        private int _pos;
+        private fixed char _default[DefaultBufferSize];
+        private const int DefaultBufferSize = 16;
```

Comment:
For stack allocation the value looks very small.
Full date format string "Thursday, January 17, 2019 5:05:29 PM" is 37 chars.
So I'd expect the const is 64 or 128 that exclude pool allocations for common formatting scenarios.

Reply:
It is very small- but it is allocated every time (whether or not you pass in your own buffer). I picked that number as that is what StringBuilder creates as a default size. Trying to get apples-to-apples here and experiment with different approaches.

I'm considering how to potentially cache what we get out of the array pool. Keeping a thread local perhaps with the largest buffer we've used (up to some max) and pulling that size by default from the pool when we grow (somewhat along the lines of what the StringBuilderCache does). I'm experimenting right now with that as copying the data on growth is a non-trivial cost.

Reply:
> I'm experimenting right now with that as copying the data on growth is a non-trivial cost.

What formatting we want to speed up? I'd collect such common formatting patterns (like datetime and other frequently used types) and measure allocations. I guess we get the const in 64-128 range. Then fallback to arraypool.